cpp/src/stream_compaction/copy_if.cuh

```diff
@@ -145,7 +145,7 @@ __global__ void scatter_kernel(T* __restrict__ output_data,
       // use an atomicOr, because these are where other blocks may overlap.
 
       constexpr int num_warps = block_size / warp_size;
-      const int last_warp = block_sum / warp_size;
+      const int last_warp = min(num_warps, (block_offset + block_sum) / warp_size);
```

Comment:
If `last_warp` refers to the last warp in the block to write to the output, then this doesn't seem quite right to me.

When `block_offset > block_size`, then `last_warp` will _always_ be equal to `num_warps`. 

For example, in a `drop_nulls`, if you had an input that looked like:
```
{ {#(block_size+1) valid values}, {#(block_size-1) NULL values}
```

Thread Block 0 will process `block_size` valid values.

Thread Block 1 will process 1 valid value, and `block_size-1` null values. `block_offset = 1024` and `block_sum = 1`, therefore, `last_warp == num_warps`, but it should actually be `0`.

Reply:
Good catch. I actually needed to account for the misalignment of the block_offset with respect to warp_size.  So the correct calculation is (I believe):

```
last_warp = min(num_warps, (block_sum + block_offset % warp_size) / warp_size);
```

Reply:
This is a tricky one to think through all the corner cases, but I think I found at least one more broken case:
```
// assume the following:
// block_offset == 1024
// block_sum == 32
// we expect warp0 in block0 to be the only participant, i.e., last_warp == 0
// however,
last_warp = (block_sum + block_offset % warp_size) / warp_size
last_warp = (32 + 1024 % 32) / 32 == 1
```

So in this example, we'd expect `last_warp` to be 0, but it ends up as `1`.

Reply:
Trying to think about how to update the calculation.

Going back to something closer to what you originally had:
```
last_warp = (block_sum == 0) ? 0 : (block_sum - 1) / warp_size;
```

I think the `last_warp` is independent of the `block_offset` and only a function of `block_sum`. As such, I think the only two edge cases we need to worry about are:
1. `block_sum == 0`
   - Need to check for this because of how 2. is solved
   - This is handled by first checking if `block_sum == 0` and if so, setting `last_warp` to 0
2. `block_sum = block_size`
   - This is handled by doing `block_sum - 1` to make sure that we always round down, even if `block_sum == block_size`

Reply:
Going off of your second solution, you could accomplish the same thing by doing something like:
```
last_warp = min(num_warps - 1, block_sum / warp_size);
```

Reply:
If a warp is active but has a 0 valid_mask then it's a nop, so while not perfectly efficient, I don't think the code is incorrect, right?

Reply:
last_warp is NOT independent of the block_offset -- thinking that is what caused the original bug this PR is fixing.   Example: if `block_offset` is 3 and `block_sum` is 30 the original calculation (and your modified calculation) will miss a bit. In this case, `last_warp` should be 1, not 0.

Reply:
I think I fixed it 4realz this time. 
```
// account for partial blocks with non-warp-aligned offsets
const int last_index = block_sum + (block_offset % warp_size) - 1;
const int last_warp = min(num_warps, last_index / warp_size);
```

The key is that the last warp is simply based on the highest index that was written in the phase before the __syncthreads. That last index is the number of true bits in the thread block plus that `aligned_offset`, but since it's an index rather than a count, you subtract one.

Reply:
Hope you can approve now. :)

Reply:
Aha, I understand the dependence on `block_offset` now. The new calculation looks great. Much more intuitive that way too.

Reply:
Yeah, once I had the insight that it is based on an index the off by one error became clear.